lib/elixir/lib/module/types/infer.ex

```diff
@@ -167,86 +188,396 @@ defmodule Module.Types.Infer do
+    :is_number,
+    :is_pid,
+    :is_port,
+    :is_reference,
+    :is_tuple
+  ]
+
   @doc """
   Refines the type variables in the typing context using type check guards
   such as `is_integer/1`.
   """
-  def of_guard(guard, context) do
-    expr_stack(guard, context, fn context ->
-      # Flatten `foo and bar` to list of type constraints
-      reduce_ok(flatten_binary_op(:and, guard), context, fn guard, context ->
-        # Flatten `foo or bar` to later build a union
-        union = Enum.map(flatten_binary_op(:or, guard), &type_guard/1)
-
-        # Only use group of guards using a single variable
-        with {:ok, args, guard_types} <- unzip_ok(union),
-             [_arg] <- Enum.uniq(Enum.map(args, &remove_meta/1)) do
-          expr_stack(guard, context, fn context ->
-            union = to_union(guard_types, context)
-            unify_guard(hd(args), union, context)
+  def of_guard({{:., _, [:erlang, :andalso]}, _, [left, right]} = expr, stack, context) do
+    stack = push_expr_stack(expr, stack)
+    fresh_context = fresh_context(context)
+
+    with {:ok, left_type, left_context} <- of_guard(left, stack, fresh_context),
+         {:ok, right_type, right_context} <- of_guard(right, stack, fresh_context),
+         {:ok, context} <- merge_context_and(context, stack, left_context, right_context),
+         {:ok, _, context} <- unify(left_type, :boolean, stack, context),
+         {:ok, _, context} <- unify(right_type, :boolean, stack, context),
+         do: {:ok, :boolean, context}
+  end
+
+  def of_guard({{:., _, [:erlang, :orelse]}, _, [left, right]} = expr, stack, context) do
+    stack = push_expr_stack(expr, stack)
+    fresh_context = fresh_context(context)
+
+    with {:ok, left_type, left_context} <- of_guard(left, stack, fresh_context),
+         {:ok, right_type, right_context} <- of_guard(right, stack, fresh_context),
+         {:ok, context} <- merge_context_or(context, stack, left_context, right_context),
+         {:ok, _, context} <- unify(left_type, :boolean, stack, context),
+         {:ok, _, context} <- unify(right_type, :boolean, stack, context),
+         do: {:ok, :boolean, context}
+  end
+
+  def of_guard({{:., _, [:erlang, :not]}, _, [arg]} = expr, stack, context) do
+    stack = push_expr_stack(expr, stack)
+    fresh_context = fresh_context(context)
+
+    with {:ok, type, arg_context} <- of_guard(arg, stack, fresh_context),
+         arg_context = invert_types(stack, arg_context),
+         {:ok, context} <- unify_new_types(context, stack, arg_context),
+         {:ok, _, context} <- unify(type, :boolean, stack, context) do
+      guard_sources = merge_guard_sources([context.guard_sources, arg_context.guard_sources])
+      {:ok, :boolean, %{context | guard_sources: guard_sources}}
+    end
+  end
+
+  def of_guard({{:., _, [:erlang, guard]}, _, args} = expr, stack, context) do
+    stack = push_expr_stack(expr, stack)
+    {param_types, return_type} = :maps.get({guard, length(args)}, @guards)
```

Comment:
We should probably move this to a function called `guard_types(...)` that receives the name and args and returns the types. It will probably be better handled by the compiler than injecting a large map and also easier to reuse. :)